distributed/shuffle/tests/test_shuffle.py

```diff
@@ -1481,8 +1395,300 @@ async def receive(self, data: list[bytes]) -> None:
         loop=loop_in_thread,
     )
     try:
-        await sB.add_partition(dfs[0])
+        await sB.add_partition(dfs[0], 0)
         with pytest.raises(RuntimeError, match="Error during receive"):
             await sB.barrier()
     finally:
         await asyncio.gather(*[s.close() for s in [sA, sB]])
+
+
+from distributed.worker import DEFAULT_EXTENSIONS
+
+
+class BlockedShuffleReceiveShuffleWorkerExtension(ShuffleWorkerExtension):
+    def __init__(self, worker: Worker) -> None:
+        super().__init__(worker)
+        self.in_shuffle_receive = asyncio.Event()
+        self.block_shuffle_receive = asyncio.Event()
+
+    async def shuffle_receive(self, *args: Any, **kwargs: Any) -> None:
+        self.in_shuffle_receive.set()
+        await self.block_shuffle_receive.wait()
+        return await super().shuffle_receive(*args, **kwargs)
+
+
+@pytest.mark.parametrize("wait_until_forgotten", [True, False])
+@mock.patch.dict(
+    DEFAULT_EXTENSIONS,
+    {"shuffle": BlockedShuffleReceiveShuffleWorkerExtension},
+)
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_deduplicate_stale_transfer(c, s, a, b, wait_until_forgotten):
```

Comment:
FYI this is the issue we ran into earlier https://github.com/dask/dask/issues/9888